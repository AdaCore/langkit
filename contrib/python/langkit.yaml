lkt_spec:
  entry_point: nodes.lkt
  source_dirs: [.]
library:
  language_name: Python
warnings:
  undocumented-nodes: false
