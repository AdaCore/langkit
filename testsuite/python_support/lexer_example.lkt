lexer foo_lexer {

    @trivia()
    whitespace <- p"[ \n\r\t]+"

    family alphanumericals {
        def <- "def"
        var <- "var"
    }

    family alphanumericals {
        error <- "error"
        example <- "example"
        null_tok <- "null"
    }


    family punctuation {
        comma <- ","
        dot <- "."
        semicolon <- ";"
        l_par <- "("
        r_par <- ")"
        l_brace <- "{"
        r_brace <- "}"
        equal <- "="
        plus <- "+"
    }

    minus <- "-"
    less_than <- "<"
    less_than_or_equal <- "<="
    greater_than <- ">"
    greater_than_or_equal <- ">="

    @unparse_spacing(with=alphanumericals)
    family alphanumericals {
        number <- p"[0-9]+|0x[0-9a-f]+"
        @symbol()
        identifier <- p"[a-zA-Z_][a-zA-Z0-9_]*"
    }

    string <- p"\"[^\"]*\""

    family comments {
        @unparse_newline_after
        @trivia()
        comment <- p"#(.?)+"
    }

}
