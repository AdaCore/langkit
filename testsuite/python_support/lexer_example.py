from __future__ import absolute_import, division, print_function

from langkit.lexer import (Lexer, LexerToken, Literal, Pattern, TokenFamily,
                           WithSymbol, WithText, WithTrivia)


class Token(LexerToken):
    Def = WithText()
    Var = WithText()
    Error = WithText()
    Example = WithText()
    Null = WithText()

    Comma = WithText()
    Dot = WithText()
    Semicolon = WithText()
    LPar = WithText()
    RPar = WithText()
    LBrace = WithText()
    RBrace = WithText()
    Equal = WithText()
    Plus = WithText()
    Minus = WithText()
    LessThan = WithText()

    Number = WithText()
    Identifier = WithSymbol()
    String = WithText()

    Comment = WithTrivia()
    Whitespace = WithTrivia()

    Alphanumericals = TokenFamily(Def, Error, Example, Null, Number,
                                  Identifier)
    Punctuation = TokenFamily(Comma, Dot, Semicolon, LPar, RPar, LBrace,
                              RBrace, Equal, Plus)
    Comments = TokenFamily(Comment)


foo_lexer = Lexer(Token)
foo_lexer.add_rules(
    (Pattern(r'[ \n\r\t]+'), Token.Whitespace),

    (Literal('def'),         Token.Def),
    (Literal('var'),         Token.Var),
    (Literal('error'),       Token.Error),
    (Literal('example'),     Token.Example),
    (Literal('null'),        Token.Null),

    (Literal(','),           Token.Comma),
    (Literal('.'),           Token.Dot),
    (Literal(';'),           Token.Semicolon),
    (Literal('('),           Token.LPar),
    (Literal(')'),           Token.RPar),
    (Literal('{'),           Token.LBrace),
    (Literal('}'),           Token.RBrace),
    (Literal('='),           Token.Equal),
    (Literal('+'),           Token.Plus),
    (Literal('-'),           Token.Minus),
    (Literal('<'),           Token.LessThan),

    (Pattern('[0-9]+'),      Token.Number),
    (Pattern('[a-zA-Z_][a-zA-Z0-9_]*'), Token.Identifier),
    (Pattern(r'\"[^\"]*\"'), Token.String),

    (Pattern('#(.?)+'), Token.Comment),
)
foo_lexer.add_spacing((Token.Alphanumericals, Token.Alphanumericals))
foo_lexer.add_newline_after(Token.Comment)
