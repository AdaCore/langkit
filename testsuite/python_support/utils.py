from __future__ import absolute_import, division, print_function

import os
import os.path
import shutil
import subprocess
import sys

import langkit
import langkit.compile_context
from langkit.compile_context import CompileCtx
from langkit.diagnostics import DiagnosticError, WarningSet
from langkit.libmanage import ManageScript

from testsuite_support.valgrind import valgrind_cmd


default_warning_set = WarningSet()

# We don't want to be forced to provide dummy docs for nodes and public
# properties in testcases.
default_warning_set.disable(WarningSet.undocumented_nodes)
default_warning_set.disable(WarningSet.undocumented_public_properties)

pretty_print = bool(int(os.environ.get('LANGKIT_PRETTY_PRINT', '0')))

project_template = """
with "libfoolang";

project Gen is
    for Languages use ("Ada");
    for Source_Dirs use (".");
    for Object_Dir use "obj";
    for Main use ({main_sources});

    package Compiler is
        for Default_Switches ("Ada") use
          ("-g", "-O0", "-gnata", "-gnatwae", "-gnatyg");
    end Compiler;
end Gen;
"""


# Determine where to find the root directory for Langkit sources
langkit_root = os.environ.get('LANGKIT_ROOT_DIR')
if not langkit_root:
    test_dir = os.path.dirname(os.path.abspath(__file__))
    testsuite_dir = os.path.dirname(test_dir)
    langkit_root = os.path.dirname(testsuite_dir)


def prepare_context(grammar, lexer=None, warning_set=default_warning_set,
                    symbol_canonicalizer=None):
    """
    Create a compile context and prepare the build directory for code
    generation.

    :param langkit.parsers.Grammar grammar: The language grammar to use for
        this context.

    :param langkit.lexer.Lexer lexer: The language lexer to use for this
        context.

    :param WarningSet warning_set: Set of warnings to emit.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    # Have a clean build directory
    if os.path.exists('build'):
        shutil.rmtree('build')
    os.mkdir('build')

    # Try to emit code
    ctx = CompileCtx(lang_name='Foo', lexer=lexer, grammar=grammar,
                     symbol_canonicalizer=symbol_canonicalizer)
    ctx.warnings = warning_set
    ctx.pretty_print = pretty_print

    return ctx


def emit_and_print_errors(grammar, lexer=None,
                          warning_set=default_warning_set,
                          generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for CTX. Return whether this was successful.

    :param langkit.parsers.Grammar grammar_fn: The language grammar to use.

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        Use `lexer_example.foo_lexer` if left to None.

    :param WarningSet warning_set: Set of warnings to emit.

    :param bool generate_unparser: Whether to generate unparser.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.

    :rtype: bool
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    try:
        ctx = prepare_context(grammar, lexer, warning_set,
                              symbol_canonicalizer=symbol_canonicalizer)
        ctx.emit('build', generate_unparser=generate_unparser)
        # ... and tell about how it went
    except DiagnosticError:
        # If there is a diagnostic error, don't say anything, the diagnostics
        # are enough.
        return False
    else:
        print('Code generation was successful')
        return True
    finally:
        lexer._dfa_code = None
        langkit.reset()


def build(grammar, lexer=None, warning_set=default_warning_set):
    """
    Shortcut for `build_and_run` to only build.
    """
    build_and_run(grammar, lexer=lexer, warning_set=warning_set)


def build_and_run(grammar, py_script=None, ada_main=None, lexer=None,
                  ocaml_main=None, warning_set=default_warning_set,
                  generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for `ctx` and build the generated library. Then,
    execute the provided scripts/programs, if any.

    An exception is raised if any step fails (the script must return code 0).

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        See emit_and_print_errors.
    :param None|str py_script: If not None, name of the Python script to run
        with the built library available.
    :param None|str|list[str] ada_main: If not None, list of name of main
        source files for Ada programs to build and run with the generated
        library. If the input is a single string, consider it's a single mail
        source file.
    :param WarningSet warning_set: Set of warnings to emit.
    :param bool generate_unparser: Whether to generate unparser.
    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    ctx = prepare_context(grammar, lexer, warning_set,
                          symbol_canonicalizer=symbol_canonicalizer)

    class Manage(ManageScript):
        def create_context(self, args):
            return ctx

    m = Manage(override_lang_source_dir=False)

    extensions_dir = os.path.abspath('extensions')
    if os.path.isdir(extensions_dir):
        ctx.extensions_dir = extensions_dir

    # First build the library. Forward all test.py's arguments to the libmanage
    # call so that manual testcase runs can pass "-g", for instance.
    argv = sys.argv[1:] + ['--full-error-traces', '-vnone', 'make']
    for w in WarningSet.available_warnings:
        argv.append('-{}{}'.format('W' if w in warning_set else 'w', w.name))
    if not pretty_print:
        argv.append('--no-pretty-print')
    if generate_unparser:
        argv.append('--generate-unparser')
    m.run(argv)

    # Flush stdout and stderr, so that diagnostics appear deterministically
    # before the script/program output.
    sys.stdout.flush()
    sys.stderr.flush()

    # Write a "setenv" script to make developper investigation convenient
    with open('setenv.sh', 'w') as f:
        m.write_setenv(f)

    env = m.derived_env()

    def run(*argv):
        subprocess.check_call(argv, env=env)

    if py_script is not None:
        # Run the Python script. Note that in order to use the generated
        # library, we have to use the special Python interpreter the testsuite
        # provides us. See the corresponding code in
        # testuite_support/python_driver.py.
        python_interpreter = os.environ['PYTHON_INTERPRETER']
        run(python_interpreter, py_script)

    if ada_main is not None:
        if isinstance(ada_main, str):
            ada_main = [ada_main]

        # Generate a project file to build the given Ada main and then run
        # the program.
        with open('gen.gpr', 'w') as f:
            f.write(project_template.format(
                main_sources=', '.join('"{}"'.format(m) for m in ada_main)
            ))
        run('gprbuild', '-Pgen', '-q', '-p',
            '-XLIBRARY_TYPE=relocatable', '-XXMLADA_BUILD=relocatable')

        for i, m in enumerate(ada_main):
            assert m.endswith('.adb')
            if i > 0:
                print('')
            if len(ada_main) > 1:
                print('== {} =='.format(m))
            sys.stdout.flush()
            argv = [os.path.join('obj', m[:-4])]
            if os.environ.get('VALGRIND_ENABLED'):
                argv = valgrind_cmd(argv)
            run(*argv)

    if ocaml_main is not None:
        run('dune', 'exec', '--display', 'quiet', '--root', '.',
            './{}.exe'.format(ocaml_main))


def add_gpr_path(dirname):
    """
    Prepend the given directory name to the ``GPR_PROJECT_PATH`` environment
    variable.
    """
    old_path = os.environ.get('GPR_PROJECT_PATH')
    os.environ['GPR_PROJECT_PATH'] = (
        '{}{}{}'.format(dirname, os.path.pathsep, old_path)
        if old_path else dirname)
