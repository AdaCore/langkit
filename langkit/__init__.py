def reset():
    """
    Reset global state in Langkit.

    TODO: this is a hack to workaround another hack. At some point in the
    future, we should get rid of this global state in Langkit.
    """

    import langkit.compile_context
    import langkit.compiled_types
    import langkit.expressions.base
    import langkit.lexer
    import langkit.utils

    langkit.utils.reset_memoized()

    langkit.compiled_types.CompiledTypeRepo.reset()
    langkit.compiled_types.create_builtin_types()

    langkit.expressions.base.Entity.unfreeze()
    langkit.expressions.base.Self.unfreeze()

    langkit.lexer.LexerToken.reset()

    langkit.compile_context.compile_ctx = None
