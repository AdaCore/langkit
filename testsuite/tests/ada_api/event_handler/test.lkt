# Check that the file reader APIs work as expected

import lexer_example

@with_lexer(foo_lexer)
grammar foo_grammar {
    @main_rule main_rule <- list+(Example("example"))
}

@abstract
class FooNode implements Node[FooNode] {
    @exported
    @external()
    fun trigger_unit_requested(name: Symbol, found: Bool, error: Bool): Bool
}

class Example: FooNode implements TokenNode {
}
