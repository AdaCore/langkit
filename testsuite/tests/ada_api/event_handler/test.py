"""
Check that the file reader APIs work as expected.
"""

from langkit.compiled_types import T
from langkit.dsl import ASTNode
from langkit.expressions import langkit_property

from utils import build_and_run


class FooNode(ASTNode):

    @langkit_property(
        return_type=T.Bool,
        external=True,
        uses_entity_info=False,
        uses_envs=False,
        public=True
    )
    def trigger_unit_requested(name=T.Symbol, found=T.Bool, error=T.Bool):
        pass


class Example(FooNode):
    token_node = True


build_and_run(
    lkt_file="expected_concrete_syntax.lkt",
    ada_main="main.adb",
    types_from_lkt=True,
)

print("Done")
