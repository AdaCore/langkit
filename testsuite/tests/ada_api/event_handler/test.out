main.adb: Starting...

== unit parsed ==

MyEH: Unit_Parsed_Callback
  Unit: <AnalysisUnit main.txt>
  Reparsed: FALSE

MyEH: Unit_Parsed_Callback
  Unit: <AnalysisUnit main.txt>
  Reparsed: TRUE

== unit requested ==

MyEH: Unit_Requested_Callback
  Name: foo_1
  From: <AnalysisUnit main.txt>
  Found: TRUE
  Is_Not_Found_Error: FALSE

MyEH: Unit_Requested_Callback
  Name: foo_2
  From: <AnalysisUnit main.txt>
  Found: FALSE
  Is_Not_Found_Error: FALSE

MyEH: Unit_Requested_Callback
  Name: foo_3
  From: <AnalysisUnit main.txt>
  Found: FALSE
  Is_Not_Found_Error: TRUE

main.adb: Done.
MyEH: Release

Done
